### DISCUSSION TITLE
Custom Rulesets Directory

### DISCUSSION THREAD

**Discussion by peppy** [ppy/osu#13096] [Show and tell]
Until we have a more defined method of centrally tracking custom rulesets (see https://github.com/ppy/osu/issues/11736), it would be appreciated if anyone working on a custom ruleset could link it (along with any details, shortcomings, etc.) in this issue. Github repository links are preferred (one comment per ruleset).

This will help us better manage breaking changes and reach out if we need further information.

Template:

```
**Name:** <name of project>
**URL:** <github url>
**Status:** <early development/wip/playable/etc>
```

Please feel free to edit your post with changelog-style updates when forward progress is made.

As an aside, a listing of dependent projects can be found [here](https://github.com/ppy/osu/network/dependents?package_id=UGFja2FnZS01NjE3NDI5MDQ%3D), which generally includes custom rulesets.

Please only post rulesets which are in a playable state.

**Comment by andy840119** [comment]
![image](https://user-images.githubusercontent.com/9100368/198080609-cb9141d1-1e9f-4beb-8ea6-420049212ee4.png)
![image](https://user-images.githubusercontent.com/9100368/198080737-5e33dfc9-0b8b-4a12-9784-d4037f16045a.png)
Name: karaoke!
URL: https://github.com/karaoke-dev/karaoke
Video: https://www.youtube.com/watch?v=Ui9mej5fhc0
Blog: https://blog.karaoke.dev/
Discord: https://discord.gg/TBTVaJW
Status: working back to the editor again.
.
Sample beatmap can download from [here](https://github.com/karaoke-dev/sample-beatmap/releases)
.
Note : 
Working on implementing full-editor for able to make karaoke beatmap from zero, so the gameplay feature might not have any update recently.
If the ruleset cause crash, please try to update to the latest version, or report in [Discord channel](https://discord.gg/TBTVaJW)

**Reply by andy840119** [reply]
2022.07 - start making the reference lyric mode for syncing the properties for those similar lyrics, see https://github.com/karaoke-dev/karaoke/issues/1444

**Reply by andy840119** [reply]
2022.10 - Start implementing the new layout in the lyric editor, see https://github.com/karaoke-dev/karaoke/issues/1594

**Reply by andy840119** [reply]
2022.12 - Start implement the classic stage editor, for able to let mapper edit the lyric position and start/end time. see https://github.com/karaoke-dev/karaoke/issues/1765

**Reply by andy840119** [reply]
2024.04 - Damn why there's no new ruleset last year?

**Comment by jorolf** [comment]
Name: pippidon
URL: https://github.com/jorolf/pippidon
Status: finished(?)

This is a barebones implementation of a ruleset which I only update rarely

**Comment by naoei** [comment]
Name: tau // τ
URL: https://github.com/naoei/tau
Status: Finished

Not my original work, original idea was from Pizzapip.

**Comment by EVAST9919** [comment]
Name: bosu!
URL: https://github.com/EVAST9919/bosu
Status: usable (but buggy due to amout of HitObjects)

Idea from I wanna be the Boshy game.

**Comment by LumpBloom7** [comment]
Name: sentakki
URL: https://github.com/LumpBloom7/sentakki
Status: Playable, still maintained

sentakki is a recreation of maimaiDX within osu! that attempts to match the mechanics and spirit of the original game. 

https://github.com/user-attachments/assets/94ab36be-36c7-4fca-92c9-548d48081914

**Reply by LumpBloom7** [reply]
2021.602.0: TOUCH notes (the blue one) are now much more enjoyable, and is now part of beatmap conversion.

<Insert forgotten versions here>

2023.924.0: A revamped beatmap converter has been introduced, providing more playable maps, among many other things. [Check out the release for more details](https://github.com/LumpBloom7/sentakki/releases/tag/2023.924.0)

**Comment by swoolcock** [comment]
Name: Rush!
URL: https://github.com/swoolcock/rush
Status: Playable, WIP

Loosely based on Muse Dash.

Edit: Prerelease now available for download.

**Comment by Flutterish** [comment]
Name: Hitokori ( Fire and Ice )
URL: https://github.com/Flutterish/Hitokori
Status: Playable

Based on `A dance of Fire and Ice`

**Comment by GSculerlor** [comment]
Name: Cytosu
URL: https://github.com/GSculerlor/Cytosu
Status: Playable. Work-in-progress, planned a lot of updates in the future.
Demo: https://www.youtube.com/watch?v=m--VLS8FCfQ

Just like its name, this ruleset based on Cytus
![osu_2021-04-29_00-14-41](https://user-images.githubusercontent.com/6506864/116445325-fa626f00-a87f-11eb-89b3-528a5c582cdc.png)

**Comment by EVAST9919** [comment]
**Name:** Swing
**URL**: https://github.com/EVAST9919/lazer-swing
**Status:** WIP (but most features have been done)
![osu_2020-08-17_17-30-45](https://user-images.githubusercontent.com/22874522/90407938-c8252e80-e0af-11ea-9718-424795f5d889.png)

**Comment by LumpBloom7** [comment]
Name: Hishigata
URL: https://github.com/LumpBloom7/hishigata
Status: Playable, with touch support on Android and iOS
Video: https://www.youtube.com/watch?v=IZp-g8tw7LU
![image](https://user-images.githubusercontent.com/12001167/96448414-a2cdb180-1213-11eb-8df0-4f35cd885473.png)

**Comment by OrangeSniper** [comment]
Name: swipe
URL: https://github.com/OrangeSniper/swipe
Status: very early, non-functional

**Comment by bdjeffyp** [comment]
Name: osu!theater
URL: https://github.com/bdjeffyp/theater-osu
Status: very early, non-functional

A game mode based on the Theatrhythm games. The plan is to support both dual input (arcade) and single input (3DS/iOS) versions of the games.

**Comment by goodtrailer** [comment]
**Name:** soyokaze!
**URL:** https://github.com/goodtrailer/soyokaze
**Status:** Finished
**Video:** https://youtu.be/hWjG0W7EiAE

A ruleset mimicking Genshin Impact's Ballads of Breeze mini-game with some extra niceties like skinning, SR calculation, statistics, etc.

![vlcsnap-2021-04-15-12h30m19s849](https://user-images.githubusercontent.com/46171121/114927732-ac149100-9de6-11eb-9fb8-5653a5e2b10d.png)

**Comment by Skyfrei** [comment]
**Name:** osu!gita (osu!ぎた)
**URL:** https://github.com/klaviotarka/osu-gita
**Status:** early development

Game mode in osu that will allow players to play via their guitars.

**Comment by NekomusuMae** [comment]
Can we place rules on when a entry can be added because at the moment, an entry can be added the moment it has a URL to link to. I'm not speaking on osu!gita, I'm talking about the multiple "rulesets" on this issue page that were published to GitHub, entered here, and then never updated even though their code is only test code. Feel free to call me out if I'm speaking out of line here but I feel we should apply a minimal level of progress before allowing an entry.

**Comment by peppy** [comment]
I've added an extra line to the opening post saying rulesets must be playable, and hidden any which aren't.

**Comment by EVAST9919** [comment]
**Name:** Touhosu
**URL:** https://github.com/EVAST9919/touhosu
**Status:** playable
![osu_2021-05-13_01-25-55](https://user-images.githubusercontent.com/22874522/118051805-4c2fec80-b38a-11eb-8137-6763f1d33f36.png)

**Comment by Flutterish** [comment]
**Name:** Solosu
**URL:** https://github.com/Flutterish/Solosu
**Status:** Playable

This ruleset is vaguely based on Solous and apparently Audiosurf
[Video](https://www.youtube.com/watch?v=iy0wMh4Ii4k) (not up to date as I completely removed tapping)

**Comment by EVAST9919** [comment]
**Name:** Sandbox
**URL:** https://github.com/EVAST9919/lazer-sandbox
**Status:** WIP, Playable

Not really a ruleset, but a set of minigames/other stuff. Currently contains:
*  2048 game
* FlappyDon (moved from framework templates)
* music visualizer (moved from m-vis, thus replacing it)

![osu_2021-06-14_05-39-39](https://user-images.githubusercontent.com/22874522/121832625-253f4000-ccd3-11eb-9aa2-f73b52619fde.png)
![2](https://user-images.githubusercontent.com/22874522/121827857-4b121800-ccc6-11eb-969a-c3123bab08c9.png)
![3](https://user-images.githubusercontent.com/22874522/121827858-4baaae80-ccc6-11eb-9f80-a370209025f4.png)
![4](https://user-images.githubusercontent.com/22874522/121827860-4baaae80-ccc6-11eb-93bd-451d07eea328.png)

**Comment by Flutterish** [comment]
Removed

**Reply by Loreos7** [reply]
OMG this is the hot one

**Reply by bdach** [reply]
the ruleset APIs were NOT intended to add fullscreen overlays like this. you should fully expect this to be broken at some point once the ruleset API surface is locked down.

it was the long term plan to have such an overlay in place, but the pieces are not there yet. i'm not super sure jumping the gun like this is helpful. it will 99% increase support burden on our side.

**Reply by Flutterish** [reply]
I am aware. I know it will break. You can tell by the amount of reflections needed to add it (well, it's like 3 calls, but any amount of reflections is too much) and the place where it loads itself. However, the support for that is on my side.

**Reply by bdach** [reply]
no it isn't. people will come here with crash reports caused by your hasty reflection use and ask why the game crashes.

ultimately it's not my final call but i would *strongly* urge reconsidering publicising something like this.

**Reply by Flutterish** [reply]
Ah, okay, I can see that. Sorry for being too hasty with my judgement about that. Though, best I can do is ensure as good reliability as possible, so you dont get any reports like that.

**Reply by peppy** [reply]
Please don't do this.

**Comment by hwabis** [comment]
**Name:** Pump Trainer
**URL:** https://github.com/hwabis/pump-trainer
**Status:** playable

A clone of Pump It Up. The main appeal is that the beatmap converter generates realistic PIU charts that are playable with 2 feet.

**Comment by michioxd** [comment]
**Name:** space
**URL:** https://github.com/michioxd/osu-space
**Status:** playable, wip

[Sound Space (Roblox)](https://www.roblox.com/games/2677609345/Sound-Space-Rhythm-Game) ported (or clone heh...) to osu! Editor implemented.
Demo: https://www.youtube.com/watch?v=Q-KHIsqJ-5E

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/4b196a1c-7539-4529-98ce-9f489d97967c" />

**Reply by OneDEV135** [reply]
so it's basically rhythia

**Comment by harsiz** [comment]
**Name:** "vis" (SoundVis)
**URL:** https://github.com/harsiz/soundvis
Status: playable, wip

I guess it's based on Beatblock and Taiko..? Maybe
<img width="1010" height="605" alt="image" src="https://github.com/user-attachments/assets/4bb373a3-df0a-49dd-9242-6260ed0569d2" />

**Comment by QingQiz** [comment]
**Name:** bms
**URL:** https://github.com/QingQiz/BmsRuleset
**Status:** Playable

# osu! BMS Ruleset

**Play BMS-family charts natively in osu!lazer**

**osu! BMS Ruleset** is a native osu!lazer ruleset plugin for playing `.bms`, `.bme`, `.bml`, and `.pms` charts.

It brings traditional BMS gameplay, keysound-based audio, background animations, gauges, scoring, difficulty tables, and detailed performance analysis into the osu! interface.

![Native BMS gameplay in osu!lazer](https://github.com/user-attachments/assets/816fb844-e684-48d0-8c37-0217fdd2d9b2)

*Native BMS gameplay in osu!lazer*

## Highlights

- **Native BMS support**  
  Play BMS-family charts directly in osu!lazer with BPM and STOP changes, keysounds, BGA, long notes, mines, random branches, scroll changes, and other commonly used BMS features.

- **Authentic judgements and scoring**  
  Uses beatoraja-compatible timing windows with PGREAT, GREAT, GOOD, BAD, POOR, and E-POOR judgements. Scores use the traditional EX-score system alongside osu!'s normalized score display.

- **Six BMS gauges**  
  Choose between Assist Easy, Easy, Normal, Hard, EX Hard, and Hazard. Auto Gauge starts from the hardest gauge and moves down through the available tiers when necessary.

- **osu!mania skin support**  
  Use your current osu!mania skin as the foundation of the playfield, provide dedicated BMS skin settings, or use the ruleset's built-in fallback skin.

- **Clear lamps**  
  Beatmap panels display your best matching local clear status, from No Play and Failed through the standard clear types to Full Combo, Perfect, and Max.

- **Difficulty tables**  
  Import LR2/beatoraja-compatible difficulty tables from URLs or JSON files. Matching charts receive difficulty markers and are automatically organized into collections.

- **BMS gameplay options**  
  Includes Auto Scratch, Hide Scratch, Background Keysound, Mirror, lane randomization, note randomization, rotation randomization, LN/CN/HCN modes, Double Time, Half Time, and more.

## Song Selection and Difficulty Tables

![Song selection, clear lamps, and difficulty tables](https://github.com/user-attachments/assets/29b2c98c-0293-4b39-8c08-7341108476d9)

Clear lamps are displayed directly in song select and follow the selected gameplay modifiers.

LR2/beatoraja difficulty tables can be imported from a URL or local JSON file. Matching charts receive table markers in their difficulty names, and each table automatically creates a collection for browsing its charts.

Well-known tables such as Satellite, Stella, Insane BMS, Overjoy, Scramble, and Luminous are available as presets.

## Detailed Result Analysis

![BMS result analysis](https://github.com/user-attachments/assets/38f62e99-8977-47eb-a0bc-d87837714905)

The result screen provides BMS-focused analysis beyond the standard score summary:

- Gauge history for every available gauge type
- Note and judgement timelines
- Fast and late timing distribution
- Hit scatter and hit-offset graphs
- Per-key timing breakdowns
- Statistics for previously saved scores

## Visual Skin Editing

![Visual BMS skin editor](https://github.com/user-attachments/assets/c3690c49-3a00-4970-abd8-0e93220d6f11)

Arrange, resize, and configure BMS gameplay components through osu!'s visual skin editor.

Editable components include the playfield, combo counter, judgement display, gauge, song progress, BGA, chart messages, and the panel comparing your current EX score and judgement counts against your personal best and target.

## Importing Charts

1. Open osu!'s settings and find the **BMS** section.
2. Select **Import BMS files**.
3. Navigate to a folder containing BMS charts.
4. Import an individual chart, the current folder, or an entire directory recursively.
5. Wait for the imported charts to appear in song select.

![BMS import screen](https://github.com/user-attachments/assets/62ff2f6d-a74a-4114-a90f-9b71d13713ca)

*Select individual charts, import the current folder, or scan a directory recursively.*

For the complete feature matrix, technical documentation, known limitations, and source code, visit the project repository.

[GitHub Repository](https://github.com/QingQiz/BmsRuleset) | [Download Releases](https://github.com/QingQiz/BmsRuleset/releases)

**Comment by ariflan159** [comment]
Name: Trace
URL: https://github.com/ariflan159/trace
Status: early development

Based on roblox game **Trace**
Fork of **tau** ruleset

**Comment by zanthous** [comment]
**Name:** sticks
**URL:** https://github.com/zanthous/Sticks
**Status:** In a good state, map editing, some level of skinning support.

Recent work: Improve visuals of double notes, improve default map converter for more pattern variety and increase patterns that require both sticks at once. Add experimental click note which is only added to maps when a converter mod is enabled 

**Zankai Games discord**: https://discord.gg/9PdMuVHbRE

Visit the #sticks channel in discord to see custom maps made by the community

**Description:** An original rhythm game played with a controller, where the player has to hit blue notes with the left stick and red notes with the right stick by flicking the stick outward and letting it reset to a neutral position inbetween notes. Sliders with reversal and double notes are the other note types

Release youtube video covering mechanics and features: https://www.youtube.com/watch?v=2Z1Irkq7Uuw


https://github.com/user-attachments/assets/fbbb76c3-d38d-45c1-b91b-cf5870967672

**Reply by 9Magnitude** [reply]
wacca but in osu

**Reply by zanthous** [reply]
Converter now has more interesting dual stick patterns:

https://github.com/user-attachments/assets/67cad708-05b9-4356-aa01-8b8600171d5a

**Comment by sumiyuki727** [comment]
Name: O2Lazer
URL: https://github.com/sumiyuki727/O2Lazer
Status: Playable(maybe)
